pulpcore/cli/common/context.py

```diff
@@ -148,41 +148,97 @@ def call(self, operation_id: str, non_blocking: bool = False, *args: Any, **kwar
+        elif task["state"] in ["waiting", "running", "canceling"]:
+            return False
+        else:
+            raise NotImplementedError(f"Unknown task state: {task['state']}")
+
+    def _poll_task(self, task: EntityDefinition, timeout: Optional[int] = None) -> EntityDefinition:
+        if timeout is None:
+            timeout = self.timeout
+
+        while True:
+            if self._check_task_finished(task):
+                click.echo("Done.", err=True)
+                return task
+            else:
+                if self.timeout:
+                    if timeout <= 0:
+                        raise PulpNoWait(f"Waiting for task {task['pulp_href']} timed out.")
+                    timeout -= 1
+                time.sleep(1)
+                click.echo(".", nl=False, err=True)
+                task = self.api.call("tasks_read", parameters={"task_href": task["pulp_href"]})
+
     def wait_for_task(self, task: EntityDefinition) -> Any:
         """
         Wait for a task to finish and return the finished task object.
+
         Raise `click.ClickException` on timeout, background, ctrl-c, if task failed or was canceled.
         """
-        timeout: int = self.timeout
         if self.background_tasks:
             raise PulpNoWait("Not waiting for task because --background was specified.")
         task_href = task["pulp_href"]
+        try:
+            return self._poll_task(task)
+        except KeyboardInterrupt:
+            raise PulpNoWait(f"Task {task_href} sent to background.")
+
+    def wait_for_task_group(self, task_group: EntityDefinition) -> Any:
+        """
+        Wait for a task group to finish and return the finished task object.
+
+        Raise `click.ClickException` on timeout, background, ctrl-c, if tasks failed or were
+        canceled.
+        """
+        timeout: int = self.timeout
+
+        if self.background_tasks:
+            raise PulpNoWait("Not waiting for task group because --background was specified.")
         try:
             while True:
-                if task["state"] == "completed":
-                    click.echo("Done.", err=True)
-                    return task
-                elif task["state"] == "failed":
-                    raise click.ClickException(
-                        f"Task {task_href} failed: '{task['error']['description']}'"
-                    )
-                elif task["state"] == "canceled":
-                    raise click.ClickException("Task canceled")
-                elif task["state"] in ["waiting", "running", "canceling"]:
+                if task_group["all_tasks_dispatched"] is True:
+                    for task in task_group["tasks"]:
+                        task = self.api.call(
+                            "tasks_read", parameters={"task_href": task["pulp_href"]}
+                        )
+                        click.echo(f"Waiting for task {task['pulp_href']}", err=True)
+                        self._poll_task(task, timeout)
```

Comment:
This is not accounting the tasks timeout as part of the whole timeout. Maybe counting seconds is not the best approach and we should be starting to compare wall clock times.

Reply:
One question I had is whether timeout should apply to each task or the entire task group?

Reply:
That is indeed a valid question. I do not know.

Reply:
I agree completely. In the best case we can shield the user from thinking about tasks. So it's one operation that takes time.